### PR TITLE
Refactor `nvexec::reduce` implementation

### PR DESCRIPTION
**Pull request by miscco**
We plan to expand the number of supported algorithms.

A lot of the necessary infrastructure will be reusable between similar algorithms, with the main exception being the `set_value` implementation.

This PR takes a first step by abstracting most of the boilerplate into a crtp base class, so that we can hide most of the complexity there.

### CODE REVIEW

File: include/nvexec/stream/reduce.cuh [hunk 1]
```diff
@@ -17,221 +17,103 @@
...
-        }
+          do {
+            if (status = STDEXEC_DBG_ERR(cub::DeviceReduce::Reduce(
+                  d_temp_storage,
+                  temp_storage_size,
+                  first,
+                  d_out,
+                  num_items,
+                  self.fun_,
+                  value_t{},
```
Comment (senior-zero) [inline include/nvexec/stream/reduce.cuh:65]: I guess non-default binary operator doesn't make a lot of sense without an ability to pass initial value. Also, in standard reduce / fold algorithms, there's `reduce(..., init)` and `reduce(..., init, binary_op)` but no `reduce(..., binary_op)`. Do you think you could fix that while we are here?

File: include/nvexec/stream/algorithm_base.cuh [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+                requires stdexec::
+                receiver_of<Receiver, completion_signatures<Self, stdexec::env_of_t<Receiver>>>
+                friend auto
+                tag_invoke(stdexec::connect_t, Self&& self, Receiver&& rcvr) -> stream_op_state_t<
+                stdexec::__copy_cvref_t<Self, Sender>,
+                receiver_t<Receiver>,
+                Receiver> {
+                return stream_op_state<stdexec::__copy_cvref_t<Self, Sender>>(
+                ((Self&&) self).sndr_,
+                (Receiver&&) rcvr,
```
Comment (senior-zero) [inline include/nvexec/stream/algorithm_base.cuh:123]: Something seems to be wrong with the formatting.

**Review by senior-zero — APPROVED** [review]

File: include/nvexec/stream/algorithm_base.cuh (hunk 2 above)
Comment (miscco) [reply include/nvexec/stream/algorithm_base.cuh:123]: Thanks a lot, I had it disabled but reenabled it :+1:

File: include/nvexec/stream/reduce.cuh (hunk 1 above)
Comment (miscco) [reply include/nvexec/stream/reduce.cuh]: Done

File: include/nvexec/stream/algorithm_base.cuh [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+          stdexec::__copy_cvref_t<Self, Sender>,
+          Env,
+          stdexec::completion_signatures<stdexec::set_error_t(cudaError_t)>,
+          set_value_t >;
+
+      template <stdexec::__decays_to<__t> Self, stdexec::receiver Receiver>
+        requires stdexec::
+                   receiver_of<Receiver, completion_signatures<Self, stdexec::env_of_t<Receiver>>>
+      friend auto tag_invoke(stdexec::connect_t, Self&& self, Receiver&& rcvr)
+        -> stream_op_state_t< stdexec::__copy_cvref_t<Self, Sender>, receiver_t<Receiver>, Receiver> {
```
Comment (ericniebler) [inline include/nvexec/stream/algorithm_base.cuh:123]:
```suggestion
      friend auto tag_invoke(stdexec::connect_t, Self&& self, Receiver rcvr)
        -> stream_op_state_t< stdexec::__copy_cvref_t<Self, Sender>, receiver_t<Receiver>, Receiver> {
```

Either that or else you need to decay the type of the receiver.

**Review by ericniebler — CHANGES_REQUESTED** [review]

**Review by ericniebler — APPROVED** [review]

**Comment by ericniebler** [comment]
Thanks!

**Pull request merged**